app/assets/javascripts/discourse/app/components/modal/grant-badge.js

```diff
@@ -0,0 +1,80 @@
+import { action } from "@ember/object";
+import Component from "@ember/component";
+import Badge from "discourse/models/badge";
+import GrantBadgeController from "discourse/mixins/grant-badge-controller";
+import I18n from "I18n";
+import UserBadge from "discourse/models/user-badge";
+import { all } from "rsvp";
+import discourseComputed from "discourse-common/utils/decorators";
+import { extractError } from "discourse/lib/ajax-error";
+import getURL from "discourse-common/lib/get-url";
+
+export default class GrantBadgeModal extends Component.extend(
+  GrantBadgeController
+) {
+  loading = true;
+  saving = false;
+  selectedBadgeId = null;
+
+  init() {
+    this.set("loading", true);
+    all([
+      Badge.findAll(),
+      UserBadge.findByUsername(this.get("post.username")),
+    ]).then(([allBadges, userBadges]) => {
+      this.setProperties({
+        allBadges,
+        userBadges,
+        loading: false,
+      });
+    });
+    super.init(...arguments);
+    this.allBadges = [];
+    this.userBadges = [];
+  }
+
+  @discourseComputed("model.selectedPost")
+  post() {
+    return this.get("model.selectedPost");
+  }
+
+  @discourseComputed("post")
+  badgeReason(post) {
+    return getURL(post.get("url"));
```

Comment:
old code here was:

```
    const url = post.get("url");
    const protocolAndHost =
      window.location.protocol + "//" + window.location.host;

    return url.startsWith("/") ? protocolAndHost + url : url;
```

This was fairly prone to errors from changes in how we construct the full URL, and did not work on local dev since the frontend / backend asset hosts are different. `getURL` should be a better approach here.